feat(medusa): add inventory_quantity support for cart items variants in store API

bqst opened the pull request:
## Summary

**What**  — What changes are introduced in this PR?

This PR adds support for `inventory_quantity` on cart item variants in the `/store/carts/:id` endpoint by applying the existing `variants.inventory_quantity` middleware from the products endpoint to the cart routes.

**Why** — Why are these changes relevant or necessary?  

- **Feature parity**: This functionality was available in [Medusa v1](https://docs.medusajs.com/v1/api/store#carts_getcartscart) and is [documented in v2](https://docs.medusajs.com/api/store#carts_getcartsid) but not implemented ([issue #12884](https://github.com/medusajs/medusa/issues/12884))
- **Performance optimization**: Eliminates the need for separate API calls to fetch inventory data, reducing the number of requests from N+1 (cart + inventory per product) to 1
- **Better UX**: Enables real-time inventory validation in cart/checkout flows without additional round trips
- **Consistency**: Aligns cart endpoint behavior with the existing `/store/products` endpoint

**How** — How have these changes been implemented?

1. Reused the existing `variants.inventory_quantity` decorator that:
  - Calculates available inventory across sales channels
  - Aggregates stock from multiple locations
  - Only returns values for variants with `manage_inventory: true`
2. No new code needed - leveraged existing infrastructure from products endpoint

**Testing** — How have these changes been tested, or how can the reviewer test the feature?

Added comprehensive integration tests in cart.spec.ts (`GET /store/carts/[id]` > `with inventory items` section):

- ✅ Returns inventory_quantity for managed inventory variants
- ✅ Returns undefined for unmanaged inventory variants
- ✅ Correctly calculates with multiple inventory items per variant
- ✅ Aggregates inventory across multiple stock locations
- ✅ Respects sales channel associations

Tests mirror the existing product endpoint tests for consistency.

---

## Examples

Provide examples or code snippets that demonstrate how this feature works, or how it can be used in practice.  
This helps with documentation and ensures maintainers can quickly understand and verify the change.

```ts
// Fetch cart with inventory quantities included
const response = await fetch(
  '/store/carts/cart_123?fields=*items.variant,+items.variant.inventory_quantity',
  {
    headers: {
      'x-publishable-api-key': 'pk_...'
    }
  }
)

const cart = await response.json()

// Access inventory quantity for each item
cart.items.forEach(item => {
  if (item.variant.manage_inventory) {
    console.log(`Product: ${item.title}`)
    console.log(`Available stock: ${item.variant.inventory_quantity}`)
    console.log(`Cart quantity: ${item.quantity}`)
    
    // Validate stock availability
    if (item.quantity > item.variant.inventory_quantity) {
      console.warn('Insufficient stock!')
    }
  }
})
```

**Before this PR:**

```ts
// Required multiple API calls
const cart = await fetch('/store/carts/cart_123')
// Then fetch each product's inventory separately
const product1 = await fetch('/store/products&id=prod_1?fields=variants.inventory_quantity')
```

**After this PR:**

```ts
// Single API call with all data
const cart = await fetch('/store/carts/cart_123?fields=*items.variant,*items.variant.inventory_quantity')
// Inventory data directly available on cart items
```

---

## Checklist

Please ensure the following before requesting a review:

- [x] I have added a **changeset** for this PR
    - Every non-breaking change should be marked as a **patch**
    - To add a changeset, run `yarn changeset` and follow the prompts
- [x] The changes are covered by relevant **tests**
- [x] I have verified the code works as intended locally
- [x] I have linked the related issue(s) if applicable

---

## Additional Context

- Related Issue: Closes #12884 
- **Documentation**: This feature is already documented in the [v2 Store API docs ](https://docs.medusajs.com/api/store#carts_getcartsid)- this PR implements the documented behavior
- v1 Compatibility: Maintains feature parity with [Medusa v1](https://docs.medusajs.com/v1/api/store#carts_getcartscart) where `inventory_quantity` was available on cart items
- Breaking Changes: None - this is a backward-compatible addition that requires explicit field selection via query parameters
- Implementation Note: Reuses existing, battle-tested middleware from `/store/products` endpoint, ensuring consistent behavior across the API

<img width="871" height="799" alt="CleanShot 2025-11-03 at 16 10 42" src="https://github.com/user-attachments/assets/eeeff970-00a8-4eac-8940-7c7e3e49882d" />


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds sales-channel-aware `inventory_quantity` to cart item variants when requested via `fields`, refactors cart routes to pass request context, and adds integration tests.
> 
> - **Store API – Carts**:
>   - Compute and expose `items.variant.inventory_quantity` (when selected via `fields`) by enhancing `refetchCart` to strip the computed field from queries and wrap fetched variants with availability.
>   - Update routes (`/store/carts`, `/store/carts/:id`, line-items add/update/delete, promotions add/remove, shipping-methods, taxes, complete) to use `MedusaStoreRequest`/`AuthenticatedMedusaRequest` and pass `req` to `refetchCart`.
> - **Inventory/Sales Channel Middleware**:
>   - Extend `wrapVariantsWithInventoryQuantityForSalesChannel` and sales-channel validation to accept both authenticated/store requests, handle missing contexts safely, and require a single sales channel for availability.
> - **Tests**:
>   - Add integration tests verifying `inventory_quantity` for managed variants, absence for unmanaged, multi-item and multi-location aggregation, and sales-channel behavior.
> - **Changeset**:
>   - Patch releases for `@medusajs/medusa` and `integration-tests-http`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3a5563c86fddcc9bca8dddbfb8fca8834b11d9cb. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Commit:
✨ Add inventory_quantity support for cart items variants in store API

Commit:
🔧 Add changeset

bqst committed:
🔧 Add changeset

olivermrbl commented on packages/medusa/src/api/store/carts/[id]/route.ts:
```diff
@@ -37,11 +52,23 @@ export const POST = async (
 
+  const withInventoryQuantity =
+    req.queryConfig.fields.some((field) => field.includes("items.variant")) &&
+    req.queryConfig.fields.some((field) => field.includes("items.variant.inventory_quantity"))
```
nit: we don't need both checks here
```suggestion
  const withInventoryQuantity =
    req.queryConfig.fields.some((field) => field.includes("items.variant.inventory_quantity"))
```

bqst replied:
I removed the double check, but this causes an issue. When using `fields=+items.variant.inventory_quantity`, the helper filters out `items.variant.inventory_quantity` but doesn't ensure `items.variant` is fetched, resulting in `cart.items[x].variant` being undefined and causing this error:

```
TypeError: Cannot read properties of undefined (reading 'id')
at wrapVariantsWithInventoryQuantityForSalesChannel
```

Should we ensure that `fields=+items.variant.inventory_quantity` automatically adds `items.variant` ?

olivermrbl commented on packages/medusa/src/api/store/carts/[id]/route.ts:
```diff
@@ -1,25 +1,40 @@
+
+  const withInventoryQuantity =
+    req.queryConfig.fields.some((field) => field.includes("items.variant")) &&
+    req.queryConfig.fields.some((field) => field.includes("items.variant.inventory_quantity"))
```
```suggestion
  const withInventoryQuantity =
    req.queryConfig.fields.some((field) => field.includes("items.variant.inventory_quantity"))
```

olivermrbl reviewed (commented):
thanks for the contribution! wonder if we should support this across all cart routes for the sake of consistency?

bqst committed:
✨ Add inventory_quantity support across all store cart routes

bqst committed:
🐛 Fix refetchCart fieldsToFetch

bqst committed:
🔧 Update changeset

bqst commented:
unstale

NicolasGorga commented on packages/medusa/src/api/store/carts/[id]/complete/route.ts:
```diff
@@ -44,18 +44,16 @@ export const POST = async (
+          defaults: defaultStoreCartFields,
+        }
+      ).remoteQueryConfig.fields
+    , req)
```
Check what this does inside refetchCart

bqst replied:
Checked: in this error path the fields are constructed from `defaultStoreCartFields` (not `req.queryConfig.fields`), and the defaults do not include `items.variant.inventory_quantity`. So `withInventoryQuantity` was always `false` here and `req` was an unused argument. Dropped it in 87383c2c5f. If the team later wants the consumer to be able to request inventory in this branch as well, we can switch to `req.queryConfig.fields` and pass `req` back in.

NicolasGorga commented on packages/medusa/src/api/utils/middlewares/products/filter-by-valid-sales-channels.ts:
```diff
@@ -8,10 +8,10 @@ import { NextFunction } from "express"
  */
 export function transformAndValidateSalesChannelIds(
-  req: MedusaStoreRequest
+  req: AuthenticatedMedusaRequest<any, any> | MedusaStoreRequest<any, any>
```
Store requests are not necessarily authenticated. I assume you made this change to handle `publishable_key_context` being undefined, but this is not necessary, since the routes where `transformAndValidateSalesChannelIds` are all `/store` routes and these are guaranteed to have the context due to `ensurePublishableApiKeyMiddleware` middleware. If you can please revert the type change :)

bqst replied:
Reverted the type back to `MedusaStoreRequest` (and dropped the `?? {}` on `publishable_key_context`). Same for `wrapVariantsWithInventoryQuantityForSalesChannel`. Done in 87383c2c5f.

NicolasGorga commented on packages/medusa/src/api/utils/middlewares/products/filter-by-valid-sales-channels.ts:
```diff
@@ -8,10 +8,10 @@ import { NextFunction } from "express"
 ): string[] {
   const { sales_channel_ids: idsFromPublishableKey = [] } =
-    req.publishable_key_context
+    req.publishable_key_context ?? {}
```
No need as per previous comment

NicolasGorga commented on packages/medusa/src/api/utils/middlewares/products/variant-inventory-quantity.ts:
```diff
@@ -26,13 +26,13 @@ export const wrapVariantsWithTotalInventoryQuantity = async (
 
   const publishableApiKeySalesChannelIds =
-    req.publishable_key_context.sales_channel_ids ?? []
+    req.publishable_key_context?.sales_channel_ids ?? []
```
Same comment as `transformAndValidateSalesChannelIds` function above.

bqst replied:
Done in 87383c2c5f. To keep the middleware clean, `refetchCart` in `helpers.ts` accepts `MedusaStoreRequest | AuthenticatedMedusaRequest` (since `POST /store/carts` and `POST /store/carts/:id/customer` use the authenticated type) and casts to `MedusaStoreRequest` once at the wrapper call site. The middleware itself is back to `MedusaStoreRequest` only.

NicolasGorga commented on packages/medusa/src/api/store/carts/helpers.ts:
```diff
@@ -4,17 +4,29 @@ import {
+  fields: string[],
+  req?: AuthenticatedMedusaRequest<any, any> | MedusaStoreRequest<any, any>
 ) => {
+  // Check if inventory_quantity is requested
+  const withInventoryQuantity =
+    fields.some((field) => field.includes("items.variant.inventory_quantity"))
+
+  // Remove inventory_quantity from fields before fetching (it's computed, not stored)
+  const fieldsToFetch = withInventoryQuantity
+    ? fields.filter((field) => !field.includes("items.variant.inventory_quantity"))
+    : fields
```
Suggestion so we guarantee the relation is expanded when inventory is requested:

```js
// along the other imports
import { deduplicate } from "@medusajs/framework/utils"

const fieldsToFetch = withInventoryQuantity
    ? deduplicate(fields.filter((field) => !field.includes("items.variant.inventory_quantity")).push("items.variant))
    : fields
```

bqst replied:
Adopted the suggestion in 87383c2c5f, with `deduplicate` from `@medusajs/framework/utils`:

```ts
const fieldsToFetch = withInventoryQuantity
  ? deduplicate([
      ...fields.filter(
        (field) => !field.includes("items.variant.inventory_quantity")
      ),
      "items.variant",
    ])
  : fields
```

Now if a consumer requests `?fields=+items.variant.inventory_quantity` without explicitly asking for `items.variant`, the relation is still expanded and the wrapper does not crash on `undefined`.

NicolasGorga commented on .changeset/gold-snakes-shave.md:
```diff
@@ -0,0 +1,6 @@
+---
+"integration-tests-http": patch
```
We don't inlcude test packages in the changeset :)

bqst replied:
Done, dropped `integration-tests-http` from the changeset in 87383c2c5f.

bqst committed:
refactor(medusa): address PR review feedback for cart inventory_quantity

- revert AuthenticatedMedusaRequest|MedusaStoreRequest union in
  filter-by-valid-sales-channels and variant-inventory-quantity middlewares.
  Store routes always have publishable_key_context via
  ensurePublishableApiKeyMiddleware, so the union and optional access were
  unnecessary (per @NicolasGorga).
- refetchCart now uses deduplicate to guarantee items.variant is fetched
  whenever items.variant.inventory_quantity is requested, preventing the
  TypeError when the consumer asks for inventory_quantity without explicitly
  requesting items.variant (per @NicolasGorga).
- complete/route.ts error path: drop the unused req argument from
  refetchCart since the constructed defaults do not include
  items.variant.inventory_quantity, making the wrapping a no-op there.
- drop integration-tests-http from changeset; test packages are not released.

bqst committed:
Merge remote-tracking branch 'upstream/develop' into feat/add-inventory-quantity-to-store-cart

# Conflicts:
#	packages/medusa/src/api/store/carts/[id]/route.ts
#	packages/medusa/src/api/store/carts/[id]/taxes/route.ts
